cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -109,13 +115,97 @@ func (spm *KubeStaticPodPathManager) BackupManifestDir() string {
+
+	// Store the backup path in the recover list. If something goes wrong now, this component will be rolled back.
+	recoverManifests[component] = backupManifestPath
+
+	// Move the old manifest into the old-manifests directory
+	if err := pathMgr.MoveFile(currentManifestPath, backupManifestPath); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	// Move the new manifest into the manifests directory
+	if err := pathMgr.MoveFile(newManifestPath, currentManifestPath); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	fmt.Printf("[upgrade/staticpods] Moved upgraded manifest to %q and backed up old manifest to %q\n", currentManifestPath, backupManifestPath)
+	fmt.Println("[upgrade/staticpods] Waiting for the kubelet to restart the component")
+
+	// Wait for the mirror Pod hash to change; otherwise we'll run into race conditions here when the kubelet hasn't had time to
+	// notice the removal of the Static Pod, leading to a false positive below where we check that the API endpoint is healthy
+	// If we don't do this, there is a case where we remove the Static Pod manifest, kubelet is slow to react, kubeadm checks the
+	// API endpoint below of the OLD Static Pod component and proceeds quickly enough, which might lead to unexpected results.
+	if err := waiter.WaitForStaticPodControlPlaneHashChange(cfg.NodeName, component, beforePodHash); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
 
-	// This string-string map stores the component name and backup filepath (if a rollback is needed).
-	// If a rollback is needed,
+	// Wait for the static pod component to come up and register itself as a mirror pod
+	if err := waiter.WaitForPodsWithLabel("component=" + component); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	fmt.Printf("[upgrade/staticpods] Component %q upgraded successfully!\n", component)
+	return nil
+}
+
+// StaticPodControlPlane upgrades a static pod-hosted control plane
+func StaticPodControlPlane(waiter apiclient.Waiter, pathMgr StaticPodPathManager, cfg *kubeadmapi.MasterConfiguration, etcdUpgrade bool) error {
 	recoverManifests := map[string]string{}
 
+	// etcd upgrade is done prior to other control plane components
+	if etcdUpgrade {
+		// Add etcd static pod spec only if external etcd is not configured
+		if len(cfg.Etcd.Endpoints) != 0 {
+			return fmt.Errorf("external etcd cannot be upgraded with kubeadm")
+
+		}
+
+		// Need to check currently used version and version from constants, if differs then uprgade
+		runningEtcdVersion, err := getEtcdRunningVersion()
+		if err != nil {
+			return err
+		}
+		currentEtcdVersion, err := semver.Parse(runningEtcdVersion)
+		if err != nil {
+			return err
+		}
+		upgradeEtcdVersion, err := constants.GetEtcdSupportedVersion(cfg.KubernetesVersion)
+		if err != nil {
+			return err
+		}
+		newEtcdVersion, err := semver.Parse(upgradeEtcdVersion)
```

Comment:
I'd use `k8s.io/kubernetes/pkg/util/version` for this task everywhere

Reply:
done